File: aoc-2022/rust/src/day2.rs
use std::fs;

enum Hand {
    Scissors,
    Paper,
    Rock
}

use Hand::{Scissors, Rock, Paper};

fn parse_line(line: (&str, &str)) -> (Hand, Hand) {
    match line {
        ("A", "X") => (Rock, Rock),
        ("A", "Y") => (Rock, Paper),
        ("A", _) => (Rock, Scissors),
        ("B", "X") => (Paper, Rock),
        ("B", "Y") => (Paper, Paper),
        ("B", _) => (Paper, Scissors),
        (_, "X") => (Scissors, Rock),
        (_, "Y") => (Scissors, Paper),
        (_, _) => (Scissors, Scissors)
    }
}

fn battle(opponent: Hand, me: Hand) -> u32 {
    match (opponent, me) {
        (Rock, Rock) => 1 + 3,
        (Rock, Paper) => 2 + 6,
        (Rock, Scissors) => 3 + 0,
        (Paper, Rock) => 1 + 0,
        (Paper, Paper) => 2 + 3,
        (Paper, Scissors) => 3 + 6,
        (Scissors, Rock) => 1 + 6,
        (Scissors, Paper) => 2 + 0,
        (Scissors, Scissors) => 3 + 3
    }
}

fn part_one(lines: &Vec<(&str, &str)>) -> u32 {
    lines.iter()
        .map(|line| parse_line(*line))
        .fold(0, |acc, elem| acc + battle(elem.0, elem.1))
}

fn reparse_line(line: (&str, &str)) -> (Hand, Hand) {
    match line {
        ("A", "X") => (Rock, Scissors),
        ("A", "Y") => (Rock, Rock),
        ("A", _) => (Rock, Paper),
        ("B", "X") => (Paper, Rock),
        ("B", "Y") => (Paper, Paper),
        ("B", _) => (Paper, Scissors),
        (_, "X") => (Scissors, Paper),
        (_, "Y") => (Scissors, Scissors),
        (_, _) => (Scissors, Rock)
    }
}

fn part_two(lines: &Vec<(&str, &str)>) -> u32 {
    lines.iter()
        .map(|line| reparse_line(*line))
        .fold(0, |acc, elem| acc + battle(elem.0, elem.1))
}

pub fn main() {
    let lines = 
        fs::read_to_string("../input/day2.txt")
        .unwrap();
    let lines = lines
        .split("\n")
        .map(|line| line.split_whitespace().collect::<Vec<_>>())
        .map(|x| (x[0], x[1]))
        .collect::<Vec<_>>();
    println!("{:?}", part_one(&lines));
    println!("{:?}", part_two(&lines));
}

File: aoc-2022/C#/Day2.cs

namespace Program
{
    class Day2
    {
        enum Result
        {
            Loss = 0, Draw = 3, Win = 6
        }
        enum Hand
        {
            Rock = 1, Paper, Scissors
        }

        int battle(Hand opponent, Hand me)
        {
            var result = (opponent, me) switch
            {
                (Hand.Rock, Hand.Paper) => Result.Win,
                (Hand.Rock, Hand.Scissors) => Result.Loss,
                (Hand.Paper, Hand.Rock) => Result.Loss,
                (Hand.Paper, Hand.Scissors) => Result.Win,
                (Hand.Scissors, Hand.Rock) => Result.Win,
                (Hand.Scissors, Hand.Paper) => Result.Loss,
                _ => Result.Draw
            };
            var score = (int)result + (int)me;
            return score;
        }

        void Part1()
        {
            var result = File.ReadAllLines("../input/day2.txt")
                .Select(line => line.Split())
                .Select(line =>
                {
                    return (line[0], line[1]) switch
                    {
                        ("A", "X") => (Hand.Rock, Hand.Rock),
                        ("A", "Y") => (Hand.Rock, Hand.Paper),
                        ("A", _) => (Hand.Rock, Hand.Scissors),
                        ("B", "X") => (Hand.Paper, Hand.Rock),
                        ("B", "Y") => (Hand.Paper, Hand.Paper),
                        ("B", _) => (Hand.Paper, Hand.Scissors),
                        (_, "X") => (Hand.Scissors, Hand.Rock),
                        (_, "Y") => (Hand.Scissors, Hand.Paper),
                        _ => (Hand.Scissors, Hand.Scissors)
                    };
                }).Aggregate(0, (acc, elem) =>
                {
                    var score = battle(elem.Item1, elem.Item2);
                    return acc + score;
                });
            Console.WriteLine(result);
        }

        void Part2()
        {
            var result = File.ReadAllLines("../input/day2.txt")
                .Select(line => line.Split())
                .Select(line =>
                {
                    return (line[0], line[1]) switch
                    {
                        ("A", "X") => (Hand.Rock, Hand.Scissors),
                        ("A", "Y") => (Hand.Rock, Hand.Rock),
                        ("A", _) => (Hand.Rock, Hand.Paper),
                        ("B", "X") => (Hand.Paper, Hand.Rock),
                        ("B", "Y") => (Hand.Paper, Hand.Paper),
                        ("B", _) => (Hand.Paper, Hand.Scissors),
                        (_, "X") => (Hand.Scissors, Hand.Paper),
                        (_, "Y") => (Hand.Scissors, Hand.Scissors),
                        _ => (Hand.Scissors, Hand.Rock)
                    };
                }).Aggregate(0, (acc, elem) =>
                {
                    var score = battle(elem.Item1, elem.Item2);
                    return acc + score;
                });
            Console.WriteLine(result);
        }

        public Day2()
        {
            // Part1();
            Part2();
        }
    }
}